### PR TITLE
Migrate code from AddressBook 3 to Mark

### LINKED ISSUE #13: Migrate AddressBook to Mark

### PR DESCRIPTION
**Pull request by dorcastan**
The following changes have been made:
* AddressBook &rightarrow; Mark (including .json files)
* Person &rightarrow; Bookmark
    * Email &rightarrow; Url (including URL validation regex)
    * Phone &rightarrow; [deleted]
    * Address &rightarrow; Remark 
* Minor UI updates – position of URL field, icon

These changes have **not** been made:
* Make Remarks an optional field
* Add Folder (involves creating an entirely new model for Folders – might be good to change Tags to a more OOP model before adapting that model for Folders)
* Update examples to make more sense. Currently, TypicalBookmarks contains (valid) bookmarks with names like Alice, Bob and Elle.

Resolves #13

### CODE REVIEW

File: docs/tutorials/AddRemark.adoc [hunk 1]
```diff
@@ -0,0 +1,425 @@
+= Tutorial - Adding a new Command
```
Comment (openorclose) [inline docs/tutorials/AddRemark.adoc]: I think this file shouldn't be here?

File: gradle/wrapper/gradle-wrapper.properties [hunk 2]
```diff
@@ -1,5 +1,6 @@
+#Thu Sep 26 17:03:44 SGT 2019
```
Comment (Na-Nazhou) [inline gradle/wrapper/gradle-wrapper.properties]: I think we don't need to change this file.

File: docs/tutorials/AddRemark.adoc (hunk 1 above)
Comment (dorcastan) [reply docs/tutorials/AddRemark.adoc]: Hm I'm not sure why it's showing up as an addition... it was merged from our team repo :/

File: gradle/wrapper/gradle-wrapper.properties (hunk 2 above)
Comment (dorcastan) [reply gradle/wrapper/gradle-wrapper.properties]: Should I change it back? It was suggested by IntelliJ, but I'm not sure why.

File: src/test/java/seedu/mark/logic/commands/EditBookmarkDescriptorTest.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+        // different name -> returns false
+        EditBookmarkDescriptor editedAmy = new EditBookmarkDescriptorBuilder(DESC_AMY).withName(VALID_NAME_BOB).build();
+        assertFalse(DESC_AMY.equals(editedAmy));
+
+        // different url -> returns false
+        editedAmy = new EditBookmarkDescriptorBuilder(DESC_AMY).withUrl(VALID_URL_BOB).build();
+        assertFalse(DESC_AMY.equals(editedAmy));
+
+        // different address -> returns false
```
Comment (Na-Nazhou) [inline src/test/java/seedu/mark/logic/commands/EditBookmarkDescriptorTest.java]:
```suggestion
        // different remark -> returns false
```

File: src/test/java/seedu/mark/logic/parser/AddCommandParserTest.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+        // multiple names - last name accepted
+        assertParseSuccess(parser, NAME_DESC_AMY + NAME_DESC_BOB + URL_DESC_BOB
+                + REMARK_DESC_BOB + TAG_DESC_FRIEND, new AddCommand(expectedBookmark));
+
+        // multiple urls - last url accepted
+        assertParseSuccess(parser, NAME_DESC_BOB + URL_DESC_AMY + URL_DESC_BOB
+                + REMARK_DESC_BOB + TAG_DESC_FRIEND, new AddCommand(expectedBookmark));
+
+        // multiple addresses - last address accepted
```
Comment (Na-Nazhou) [inline src/test/java/seedu/mark/logic/parser/AddCommandParserTest.java]:
```suggestion
        // multiple remarks - last remark accepted
```

File: src/test/java/seedu/mark/logic/parser/AddCommandParserTest.java [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+        // missing name prefix
+        assertParseFailure(parser, VALID_NAME_BOB + URL_DESC_BOB + REMARK_DESC_BOB,
+                expectedMessage);
+
+        // missing url prefix
+        assertParseFailure(parser, NAME_DESC_BOB + VALID_URL_BOB + REMARK_DESC_BOB,
+                expectedMessage);
+
+        // missing address prefix
```
Comment (Na-Nazhou) [inline src/test/java/seedu/mark/logic/parser/AddCommandParserTest.java]:
```suggestion
        // missing remark prefix
```

File: src/test/java/seedu/mark/logic/parser/AddCommandParserTest.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    public void parse_invalidValue_failure() {
+        // invalid name
+        assertParseFailure(parser, INVALID_NAME_DESC + URL_DESC_BOB + REMARK_DESC_BOB
+                + TAG_DESC_HUSBAND + TAG_DESC_FRIEND, Name.MESSAGE_CONSTRAINTS);
+
+        // invalid url
+        assertParseFailure(parser, NAME_DESC_BOB + INVALID_URL_DESC + REMARK_DESC_BOB
+                + TAG_DESC_HUSBAND + TAG_DESC_FRIEND, Url.MESSAGE_CONSTRAINTS);
+
+        // invalid address
```
Comment (Na-Nazhou) [inline src/test/java/seedu/mark/logic/parser/AddCommandParserTest.java]:
```suggestion
        // invalid remark
```

File: src/test/java/seedu/mark/logic/parser/EditCommandParserTest.java [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+
+        // invalid prefix being parsed as preamble
+        assertParseFailure(parser, "1 i/ string", MESSAGE_INVALID_FORMAT);
+    }
+
+    @Test
+    public void parse_invalidValue_failure() {
+        assertParseFailure(parser, "1" + INVALID_NAME_DESC, Name.MESSAGE_CONSTRAINTS); // invalid name
+        assertParseFailure(parser, "1" + INVALID_URL_DESC, Url.MESSAGE_CONSTRAINTS); // invalid url
+        assertParseFailure(parser, "1" + INVALID_REMARK_DESC, Remark.MESSAGE_CONSTRAINTS); // invalid address
```
Comment (Na-Nazhou) [inline src/test/java/seedu/mark/logic/parser/EditCommandParserTest.java]:
```suggestion
        assertParseFailure(parser, "1" + INVALID_REMARK_DESC, Remark.MESSAGE_CONSTRAINTS); // invalid remark
```

File: src/test/java/seedu/mark/logic/parser/EditCommandParserTest.java [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+        EditCommand expectedCommand = new EditCommand(targetIndex, descriptor);
+        assertParseSuccess(parser, userInput, expectedCommand);
+
+        // url
+        userInput = targetIndex.getOneBased() + URL_DESC_AMY;
+        descriptor = new EditBookmarkDescriptorBuilder().withUrl(VALID_URL_AMY).build();
+        expectedCommand = new EditCommand(targetIndex, descriptor);
+        assertParseSuccess(parser, userInput, expectedCommand);
+
+        // address
```
Comment (Na-Nazhou) [inline src/test/java/seedu/mark/logic/parser/EditCommandParserTest.java]:
```suggestion
        // remark
```

File: src/test/java/seedu/mark/model/bookmark/BookmarkTest.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+        // different name -> returns false
+        Bookmark editedAlice = new BookmarkBuilder(ALICE).withName(VALID_NAME_BOB).build();
+        assertFalse(ALICE.equals(editedAlice));
+
+        // different url -> returns false
+        editedAlice = new BookmarkBuilder(ALICE).withUrl(VALID_URL_BOB).build();
+        assertFalse(ALICE.equals(editedAlice));
+
+        // different address -> returns false
```
Comment (Na-Nazhou) [inline src/test/java/seedu/mark/model/bookmark/BookmarkTest.java]:
```suggestion
        // different remark -> returns false
```

File: src/main/java/seedu/mark/model/bookmark/Url.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+            + REGEX_URL_PATH + REGEX_URL_QUERY + REGEX_URL_FRAGMENT + "/?$";
+
+    public final String value;
+
+    /**
+     * Constructs a {@code Url}.
+     *
+     * @param email A valid URL.
+     */
+    public Url(String email) {
```
Comment (Na-Nazhou) [inline src/main/java/seedu/mark/model/bookmark/Url.java]:
```suggestion
    public Url(String url) {
```

File: src/test/java/seedu/mark/logic/commands/EditBookmarkDescriptorTest.java (hunk 3 above)
Comment (Na-Nazhou) [reply src/test/java/seedu/mark/logic/commands/EditBookmarkDescriptorTest.java]: There are a few more similar instances elsewhere as well

File: src/main/java/seedu/mark/model/util/SampleDataUtil.java [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+import seedu.mark.model.bookmark.Name;
+import seedu.mark.model.bookmark.Remark;
+import seedu.mark.model.bookmark.Url;
+import seedu.mark.model.tag.Tag;
+
+/**
+ * Contains utility methods for populating {@code Mark} with sample data.
+ */
+public class SampleDataUtil {
+    public static Bookmark[] getSampleBookmarks() {
```
Comment (Na-Nazhou) [inline src/main/java/seedu/mark/model/util/SampleDataUtil.java]: In order to run the app, the URL has to be replaced with valid URL to pass the validation. After changing to valid URLs, I managed to run the app locally :)

I collected some random URLs you can just copy paste :P
```
    public static Bookmark[] getSampleBookmarks() {
        return new Bookmark[] {
            new Bookmark(new Name("Alex Yeoh"), new Url("https://www.google.com/"),
                new Remark("Blk 30 Geylang Street 29, #06-40"),
                getTagSet("friends")),
            new Bookmark(new Name("Bernice Yu"), new Url("https://www.youtube.com/"),
                new Remark("Blk 30 Lorong 3 Serangoon Gardens, #07-18"),
                getTagSet("colleagues", "friends")),
            new Bookmark(new Name("Charlotte Oliveiro"), new Url("http://www.nus.edu.sg/"),
                new Remark("Blk 11 Ang Mo Kio Street 74, #11-04"),
                getTagSet("neighbours")),
            new Bookmark(new Name("David Li"), new Url("https://en.wikipedia.org/"),
                new Remark("Blk 436 Serangoon Gardens Street 26, #16-43"),
                getTagSet("family")),
            new Bookmark(new Name("Irfan Ibrahim"), new Url("https://www.reddit.com/"),
                new Remark("Blk 47 Tampines Street 20, #17-35"),
                getTagSet("classmates")),
            new Bookmark(new Name("Roy Balakrishnan"), new Url("https://www.facebook.com/"),
                new Remark("Blk 45 Aljunied Street 85, #11-31"),
                getTagSet("colleagues"))
        };
    }
```

File: docs/tutorials/AddRemark.adoc [hunk 12]
```diff
@@ -266,12 +266,13 @@ These should be relatively simple changes.
 Without getting too deep into `fxml`, let's go on a 5 minute adventure to get some placeholder text to show up for each person.
```
[the remark this answers is not in the document]
Comment (dorcastan) [reply docs/tutorials/AddRemark.adoc]: I've reverted all changes to the text of the tutorials. These two changes to `[source, java]` actually correct wrong sidebar block formatting. They were pointed out in one of the automated checks, but I'm not sure if they are crucial for the checks to pass. Should I change it back?

**Pull request merged**